File: python/problems/leetcode/0787_cheapest_flights_within_k_stops/solution.py
"""LeetCode 787: Cheapest Flights Within K Stops."""

from __future__ import annotations


class Solution:
    def findCheapestPrice(
        self,
        n: int,
        flights: list[list[int]],
        src: int,
        dst: int,
        k: int,
    ) -> int:
        inf = 10**9
        dist = [inf] * n
        dist[src] = 0
        for _ in range(k + 1):
            nxt = dist[:]
            for u, v, w in flights:
                if dist[u] != inf:
                    nxt[v] = min(nxt[v], dist[u] + w)
            dist = nxt
        return dist[dst] if dist[dst] != inf else -1


if __name__ == "__main__":
    assert Solution().findCheapestPrice(3, [[0, 1, 100], [1, 2, 100], [0, 2, 500]], 0, 2, 1) == 200
    assert Solution().findCheapestPrice(3, [[0, 1, 100], [1, 2, 100], [0, 2, 500]], 0, 2, 0) == 500
    assert Solution().findCheapestPrice(3, [[0, 1, 100], [1, 2, 100]], 0, 2, 1) == 200
    print("0787 OK")

File: cpp/problems/leetcode/0787_cheapest_flights_within_k_stops/solution.cpp
// LeetCode 787: Cheapest Flights Within K Stops
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        const int inf = 1e9;
        vector<int> dist(n, inf);
        dist[src] = 0;
        for (int round = 0; round <= k; ++round) {
            vector<int> nxt = dist;
            for (auto& f : flights) {
                int u = f[0], v = f[1], w = f[2];
                if (dist[u] != inf) nxt[v] = min(nxt[v], dist[u] + w);
            }
            dist = move(nxt);
        }
        return dist[dst] < inf ? dist[dst] : -1;
    }
};

int main() {
    vector<vector<int>> f{{0, 1, 100}, {1, 2, 100}, {0, 2, 500}};
    assert(Solution().findCheapestPrice(3, f, 0, 2, 1) == 200);
    assert(Solution().findCheapestPrice(3, f, 0, 2, 0) == 500);
    vector<vector<int>> f2{{0, 1, 100}, {1, 2, 100}};
    assert(Solution().findCheapestPrice(3, f2, 0, 2, 1) == 200);
    cout << "0787 OK" << endl;
    return 0;
}
